File: Civitas - Ruby/pkg/Civitas_Ruby-0.0.1/lib/mazo_sorpresas.rb
# To change this license header, choose License Headers in Project Properties.
# To change this template file, choose Tools | Templates
# and open the template in the editor.
module Civitas
  class MazoSorpresa
    #Metodo privado de inicializacion
    private
    def init
      @sorpresas = []
      @cartas_especiales = []
      @barajada = false
      @usadas = 0
    end
    
    public
    def initialize(d=false)
      @debug = d
      init
      if @debug
        Diario.instance.ocurre_evento("El modo debug esta activo")
      end
    end
  
    #Inserta una carta si no esta el mazo en uso
    def alMazo(s)
      if !@barajada
      @sorpresas.push(s)
      end
    end
    
    #Sacar una carta del mazo
    def siguiente
      if ((!@barajada || @usadas == @sorpresas.length) && !@debug)  #Se baraja si se han usado todas o no se ha barajado
        @sorpresas.shuffle
        @usadas = 0
        @barajada = true
      end

      @usadas = @usadas +1
      @ultima_sorpresa = @sorpresas[0]
      @sorpresas.shift
      @sorpresas.push(@ultima_sorpresa)

      return @ultima_sorpresa
    end
    
    
    def inhabilitarCartaEspecial(sorpresa)
      if @sorpresas.include?(sorpresa)
        @sorpresas.delete[sorpresa]
        @cartas_especiales.push(sorpresa)
        Diario.instance.ocurre_evento("Se ha inhabilitado una carta especial")
      end
    end
  
    
    def habilitarCartaEspecial(sorpresa)
      if @cartas_especiales.include?(sorpresa)
        @cartas_especiales.delete[sorpresa]
        @sorpresas.push(sorpresa)
        Diario.instance.ocurre_evento("Se ha habilitado una carta especial")
      end
    end
    
    
  end
end

File: Civitas - Java/src/civitas/MazoSorpresas.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package civitas;

import java.util.ArrayList;
import java.util.*;
/**
 *
 */
public class MazoSorpresas {
    private ArrayList<Sorpresa> sorpresas;
    private boolean barajada;
    private int usadas;
    private boolean debug;
    private ArrayList<Sorpresa> cartasEspeciales;
    private Sorpresa ultimaSorpresa;
    
    private void init(){
        sorpresas = new ArrayList<>();
        cartasEspeciales = new ArrayList<>();
        barajada = false;
        usadas = 0;
    }
    
    //Constructor con parámetros
    MazoSorpresas(boolean deb){
        debug = deb;
        this.init();
        if (debug)
            Diario.getInstance().ocurreEvento("El modo debug está activo");
    }
    
    //Constructor sin parámetros
    MazoSorpresas(){
        this.init();
        debug = false;
    }
    
    //Si el mazo no se ha barajado, se añade la sorpresa al mazo
    void alMazo(Sorpresa s){
        if(!barajada)
            sorpresas.add(s);
    }
    
    //Si el mazo no ha sido barajado o si el numero de cartas usadas es igual al tamaño del mazo, se baraja
    Sorpresa siguiente(){
        if(!barajada || usadas == sorpresas.size() && !debug){
            Collections.shuffle(sorpresas);  //Sirve para barajar el mazo
            barajada = true;
            usadas=0;
        }
        usadas ++;
        ultimaSorpresa = sorpresas.get(0);
        sorpresas.remove(0);
        sorpresas.add(ultimaSorpresa);
        
        return ultimaSorpresa;    
    }
    
    //Si la sorpresa es carta especial y está en el mazo, se quita y se añade a cartasEpeciales
    void inhabilitarCartaEspecial (Sorpresa sorpresa){
        if(sorpresas.contains(sorpresa)){
            cartasEspeciales.add(sorpresa);
            sorpresas.remove(sorpresa);
            Diario.getInstance().ocurreEvento("Se ha inhabilitado una carta especial");
        }
    }
    
    //Si la sorpresa está en cartaEspecial se saca y se añade al final de la coleccion de sorpresas
    void habilitarCartaEspecial(Sorpresa sorpresa){
      if(cartasEspeciales.contains(sorpresa)){
            cartasEspeciales.remove(sorpresa);
            sorpresas.add(sorpresa);
            Diario.getInstance().ocurreEvento("Se ha habilitado una carta especial");
        }    
    }

    Sorpresa getSorpresas(int i){
        return sorpresas.get(i);
    }
    
    int getTamSorpresas(){
        return sorpresas.size();
    }
}